cmd/util/ledger/migrations/cadence_value_validation.go

```diff
@@ -0,0 +1,599 @@
+package migrations
+
+import (
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/onflow/atree"
+	"github.com/onflow/cadence"
+	"github.com/onflow/cadence/runtime"
+	"github.com/onflow/cadence/runtime/common"
+	"github.com/onflow/cadence/runtime/interpreter"
+	"github.com/rs/zerolog"
+	"go.opentelemetry.io/otel/attribute"
+
+	"github.com/onflow/flow-go/cmd/util/ledger/util"
+	"github.com/onflow/flow-go/ledger"
+)
+
+var nopMemoryGauge = util.NopMemoryGauge{}
+
+// TODO: optimize memory by reusing payloads snapshot created for migration
+func validateCadenceValues(
+	address common.Address,
+	oldPayloads []*ledger.Payload,
+	newPayloads []*ledger.Payload,
+	log zerolog.Logger,
+	verboseLogging bool,
+) error {
+	// Create all the runtime components we need for comparing Cadence values.
+	oldRuntime, err := newReadonlyStorageRuntime(oldPayloads)
+	if err != nil {
+		return fmt.Errorf("failed to create validator runtime with old payloads: %w", err)
+	}
+
+	newRuntime, err := newReadonlyStorageRuntime(newPayloads)
+	if err != nil {
+		return fmt.Errorf("failed to create validator runtime with new payloads: %w", err)
+	}
+
+	// Iterate through all domains and compare cadence values.
+	for _, domain := range domains {
```

Comment:
We can but most of the data is in one domain, so it is likely to remain a bottleneck (not much to gain by parallelism).

Reply:
Could we validate the storage map keys in parallel though?

Reply:
> Could we validate the storage map keys in parallel though?

Good question.  That would allow parallelizing within same domain but we would need to determine if updating atree read cache needs to support parallelism for this.